srml/contracts/src/lib.rs

```diff
@@ -431,11 +431,13 @@ decl_module! {
 			// can alter the balance of the caller.
 			gas::refund_unused_gas::<T>(&origin, gas_meter, imbalance);
 
-			// Dispatch every recorded call with an appropriate origin.
-			ctx.calls.into_iter().for_each(|(who, call)| {
-				let result = call.dispatch(RawOrigin::Signed(who.clone()).into());
-				Self::deposit_event(RawEvent::Dispatched(who, result.is_ok()));
-			});
+			if let Ok(_) = result {
+				// Dispatch every recorded call with an appropriate origin.
+				ctx.calls.into_iter().for_each(|(who, call)| {
```

Comment:
Is there a reason this has to be below `refund_unused_gas` and can't go in the `if` block above immediately after the events are deposited?

Reply:
Not sure. Perhaps @pepyakin or @thiolliere can elaborate?

> (However, we still need to dispatch after refunding, not before)

Reply:
I would say that as gas is not used anymore its better that the caller got its money back before processing the logic of those calls. They might want to make use of it and there is no reason for us to hold it longer

Reply:
That would be a nice comment to add to the source. :-)